src/sciline/data_graph.py

```diff
@@ -0,0 +1,205 @@
+        out._cbgraph = graph
+        return out
+
+    def copy(self: T) -> T:
+        return self._from_cyclebane(self._cbgraph.copy())
+
+    def __copy__(self: T) -> T:
+        return self.copy()
+
+    @property
+    def _graph(self) -> nx.DiGraph:
+        return self._cbgraph.graph
+
+    def _get_clean_node(self, key: Key) -> Any:
+        """Return node ready for setting value or provider."""
+        if key is NoneType:
+            raise ValueError('Key must not be None')
+        if key in self._graph:
+            self._graph.remove_edges_from(list(self._graph.in_edges(key)))
+            self._graph.nodes[key].pop('value', None)
+            self._graph.nodes[key].pop('provider', None)
+            self._graph.nodes[key].pop('reduce', None)
+        else:
+            self._graph.add_node(key)
+        return self._graph.nodes[key]
+
+    def insert(self, provider: Union[ToProvider, Provider], /) -> None:
+        """
+        Insert a callable into the graph that provides its return value.
+
+        Parameters
+        ----------
+        provider:
+            Either a callable that provides its return value. Its arguments
+            and return value must be annotated with type hints.
+            Or a ``Provider`` object that has been constructed from such a callable.
+        """
+        if not isinstance(provider, Provider):
+            provider = Provider.from_function(provider)
+        return_type = provider.deduce_key()
+        if typevars := _find_all_typevars(return_type):
+            for bound in _mapping_to_constrained(typevars):
+                self.insert(provider.bind_type_vars(bound))
+            return
+        # Trigger UnboundTypeVar error if any input typevars are not bound
+        provider = provider.bind_type_vars({})
+        self._get_clean_node(return_type)['provider'] = provider
+        for dep in provider.arg_spec.keys():
+            self._graph.add_edge(dep, return_type, key=dep)
+
+    def __setitem__(self, key: Key, value: DataGraph | Any) -> None:
+        """
+        Provide a concrete value for a type.
+
+        Parameters
+        ----------
+        key:
+            Type to provide a value for.
+        param:
+            Concrete value to provide.
+        """
+        # This is a questionable approach: Using MyGeneric[T] as a key will actually
+        # not pass mypy [valid-type] checks. What we do on our side is ok, but the
+        # calling code is not.
+        if typevars := _find_all_typevars(key):
+            for bound in _mapping_to_constrained(typevars):
+                self[_bind_free_typevars(key, bound)] = value
+            return
+
+        # TODO If key is generic, should we support multi-sink case and update all?
+        # Would imply that we need the same for __getitem__.
+        self._cbgraph[key] = (
+            value._cbgraph if isinstance(value, DataGraph) else _as_graph(key, value)
+        )
+
+    def __getitem__(self: T, key: Key) -> T:
```

Comment:
I just noticed that `__setitem__` and `__getitem__` are asymmetric:
```python
graph[key] = value
assert graph[key] != value
```
Can we avoid this somehow?

Reply:
We can avoid it be requiring wrapping the RHS in a `DataGraph`:

```
graph[key] = DataGraph.from_value(value)  # create single-node graph with value
```

The current behavior is a convenience feature for that.

Reply:
That would be too inconvenient. Esp. when we looks at the user feedback we got. But what about `add_param` or `set_param` or something similar? Then we are explicit and it should avoid confusion around the setitem syntax.

Reply:
I don't think that is convenient or obvious for users? Scientist may be familiar with `dict`, so if we can stick with that it would  be good?

Reply:
Maybe. I don't have evidence for either